pkg/fqdn/regexpmap/regexp_map.go

```diff
@@ -19,10 +19,109 @@ import (
 	"sort"
 )
 
-// lookupValueSet is a utility type. It is intended as a set of strings
-// inserted into a RegexpMap. Lookups are mapped to number of times
-// they were added to map.
-type lookupValueSet map[string]int
+// RegexpList is a utility struct that keeps an array of strings sorted by
+// length
+type RegexpList []string
+
+// NewRegexpList returns a new RegexpList, if any initialValues is in place
+// will add into the utility.
+func NewRegexpList(initialValues ...string) (result *RegexpList) {
+	sort.Slice(initialValues, func(i, j int) bool {
+		return len(initialValues[i]) < len(initialValues[j])
+	})
+
+	tmp := RegexpList(initialValues)
+	return &tmp
+}
+
+// Add function adds a new item in the list and sort the data based on the
+// length
+func (r *RegexpList) Add(val string) {
+	for _, item := range *r {
+		if item == val {
+			return
+		}
+	}
+
+	tmpData := append(*r, val)
+
+	sort.Slice(tmpData, func(i, j int) bool {
+		return len(tmpData[i]) < len(tmpData[j])
+	})
+
+	*r = tmpData
+}
+
+// Remove removes the item from the internal array and keep the data sorted
+func (r *RegexpList) Remove(val string) {
+	tmpData := []string{}
+	for _, item := range *r {
+		if item == val {
+			continue
+		}
+		tmpData = append(tmpData, item)
+	}
+	*r = tmpData
+	return
+}
+
+// Empty empties the internal array
+func (r *RegexpList) Empty() {
+	*r = []string{}
+}
+
+// Get return an array of strings ordered by string len
+func (r *RegexpList) Get() []string {
+	return *r
+}
+
+// Len returns the len of the internal array.
+func (r *RegexpList) Len() int {
+	return len(*r)
+}
+
+// refCount is a map alias that has a specific options for reference counts
+type refCount map[string]int
+
+// Increment adds one to the given key
+func (r refCount) Increment(key string) int {
+	r[key]++
+	return r[key]
+}
+
+// Decrement remove one to the given key. If the value is 0, the key will be
+// deleted.
+func (r refCount) Decrement(key string) int {
+	val, exists := r[key]
+	if !exists {
+		return 0
+	}
+
+	if val <= 0 {
+		return 0
+	}
+
+	if val == 1 {
+		delete(r, key)
+		return 0
+	}
+	r[key]--
+	return r[key]
+}
+
+// Keys return the list of keys that are in place.
+func (r refCount) Keys() []string {
+	result := []string{}
```

Comment:
I have a very slight preference on relying on `append` to create this when needed. We use this function in `LookupValues` (only there, in fact) and unless the compiler is smart (this is very possible) it will allocate an object even when no values are present. Relying on append ensures that.
Perhaps this can't happen, though? as we would clean up the refcount object in `rulesRelation` if it was empty?